enode.rkt

```diff
@@ -233,9 +236,20 @@
 (define (enode-vars en)
   (enode-cvars (pack-leader en)))
 
+(define (update-en-expr expr)
+  (if (list? expr)
+      (for/list ([sub (in-list expr)])
+        (if (enode? sub) (pack-leader sub) sub))
+      expr))
+
+
 ;; Removes duplicates from the varset of this node.
 (define (refresh-vars! en)
-  (set-enode-cvars! en (list->set (set->list (enode-cvars en)))))
+  (set-enode-cvars!
+   en
+   (list->set
+    (for/list ([cvars (enode-cvars en)])
+      (update-en-expr cvars)))))
```

Comment:
`for/set` instead of `for/list`, and also wrap the iteration over `enode-cvars` with `in-set` so it goes a tiny bit faster.